resolwe/flow/managers/base.py

```diff
@@ -87,11 +198,133 @@ def _include_environment_variables(self, program):
 
-    def run(self, data_id, script, priority='normal', run_sync=False, verbosity=1):
+    def run(self, data_id, dest_dir, argv, priority='normal', run_sync=False, verbosity=1):
         """Run process."""
         raise NotImplementedError('`run` function not implemented')
 
-    def communicate(self, run_sync=False, verbosity=1):
+    def _get_per_data_dir(self, dir_base, data_id):
+        """Return the given base directory extended with a per-data component."""
+        result = getattr(settings, 'FLOW_EXECUTOR', {}).get(dir_base, '')
+        return os.path.join(result, str(data_id))
+
+    def _prepare_data_dir(self, data_id):
+        """Prepare destination directory where the data will live."""
+        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('DATA_DIR_MODE', 0o755)
+        os.mkdir(output_path, mode=dir_mode)
+        # os.mkdir is not guaranteed to set the given mode
+        os.chmod(output_path, dir_mode)
+        return output_path
+
+    def _prepare_context(self, data_id, data_dir, runtime_dir, **kwargs):
+        """Prepare settings and constants jsons for the executor."""
+        settings_dict = dict(kwargs)
+        settings_dict.update(self._marshal_settings())
+        settings_dict['DATA_DIR'] = data_dir
+
+        # add scheduling classes
+        settings_dict['Process_sched_class'] = {
+            k: getattr(Process, k) for k in dir(Process)
+            if k.startswith('SCHEDULING_CLASS_') and isinstance(getattr(Process, k), str)
+        }
+        # add Data status constants
+        settings_dict['Data_status'] = {
+            k: getattr(Data, k) for k in dir(Data)
+            if k.startswith('STATUS_') and isinstance(getattr(Data, k), str)
+        }
+        # required plumbing settings
+        settings_dict['REDIS_CHANNEL_PAIR'] = state.MANAGER_EXECUTOR_CHANNELS
+
+        # extend the settings with whatever the executor wants
+        settings_dict.update(self.executor.make_settings(data_id))
+
+        # save the settings into a json in the data dir
+        json.dump(settings_dict, open(os.path.join(runtime_dir, 'settings.json'), 'wt'), cls=SettingsJSONifier)
+
+    def _prepare_executor(self, data_id, executor):
+        """Copy executor sources into the destination directory."""
+        import resolwe.flow.executors as executor_module
+
+        exec_dir = executor_module.__path__[0]
+        dest_dir = self._get_per_data_dir('RUNTIME_DIR', data_id)
+        module_dir = os.path.join(dest_dir, 'executors')
+        shutil.copytree(exec_dir, module_dir)
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('RUNTIME_DIR_MODE', 0o755)
+        os.chmod(dest_dir, dir_mode)
+
+        class_name = executor.split('.')
+        class_name = class_name[class_name.index('executors') + 1:]
+        return '.' + '.'.join(class_name) + '.run.FlowExecutor', dest_dir
```

Comment:
This part is also messy. Perhaps create an intermediate variable like:

```python
dest_class_path = '.{}.run.FlowExecutor'.format(class_name)
```

Also, will the '.run.FlowExecutor' part be obtained from `protocol.py` module?

Reply:
Not entirely applicable anymore. The `.run.FlowExecutor` part is now executor-internal and appears as a string literal there. The manager must supply the concrete executor's _module_ name relative to the `executors` module. Instantiation is then the `run_executor()` driver's responsibility, as is the concrete class hierarchy.

Reply:
Ok, good!